src/Type/Php/CtypeDigitFunctionTypeSpecifyingExtension.php

```diff
@@ -54,7 +56,18 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
 			]);
 		}
 
-		return $this->typeSpecifier->create($node->getArgs()[0]->value, TypeCombinator::union(...$types), $context, false, $scope);
+		$unionType = TypeCombinator::union(...$types);
+		$specifiedTypes = $this->typeSpecifier->create($exprArg, $unionType, $context, false, $scope);
+
+		if ($context->true() && $exprArg instanceof Cast) {
+			$castOriginType = TypeCombinator::intersect($scope->getType($exprArg->expr), $unionType);
+
+			$specifiedTypes = $specifiedTypes->unionWith(
+				$this->typeSpecifier->create($exprArg->expr, $castOriginType, $context, false, $scope)
```

Comment:
Are you sure you can't just use `$unionType` instead of `$castOriginType` here in the 2nd arg?

Reply:
good point. fixed.